### PR TITLE
Interface cache

### PR DESCRIPTION
**Pull request by DieMyst**

### CODE REVIEW

File: crates/test-utils/src/utils.rs [hunk 1]
```diff
@@ -1,4 +1,9 @@
+use fluence_faas::TomlFaaSNamedModuleConfig;
+use host_closure::Args;
 use ivalue_utils::IValue;
+pub use particle_modules::Hash;
+pub use particle_modules::{AddBlueprint, Dependency, ModuleRepository};
+use serde_json::Value as JValue;
```
Comment (justprosh) [inline crates/test-utils/src/utils.rs:6]: I think you should group imports like this:
- local imports
- other things from fluence org (like fce, aquamarine imports)
- other imports

File: particle-modules/src/modules.rs [hunk 2]
```diff
@@ -15,23 +15,26 @@
...
 use fluence_app_service::ModuleDescriptor;
 use host_closure::{closure, Args, Closure};
 
 use eyre::WrapErr;
 use fstrings::f;
-use parking_lot::Mutex;
+use parking_lot::{Mutex, RwLock};
 use serde::{Deserialize, Serialize};
-use serde_json::{json, Value as JValue};
+use serde_json::{json, Value as JValue, Value};
```
Comment (justprosh) [inline particle-modules/src/modules.rs:37]: Maybe leave only one `Value`

File: particle-node/tests/network_explore.rs [hunk 3]
```diff
@@ -120,13 +120,22 @@ fn get_modules() {
...
+                    (seq
+                        (call relay ("dist" "list_modules") [] modules)
+                        (fold modules m
+                            (seq
+                                (call relay ("dist" "get_module_interface") [m.$.hash!] interfaces[])
+                                (next m)
+                            )
+                        )               
+                    )
+
```
Comment (justprosh) [inline particle-node/tests/network_explore.rs:135]: redundant blank line

**Review by justprosh — APPROVED** [review]

File: particle-modules/src/modules.rs [hunk 4]
```diff
@@ -79,10 +86,20 @@ impl ModuleRepository {
 
         let modules_by_name = Arc::new(Mutex::new(modules_by_name));
 
+        let blueprints = Self::load_blueprints(blueprints_dir);
+        let mut bp_map = HashMap::new();
+        for bp in blueprints.iter() {
+            bp_map.insert(bp.id.clone(), bp.clone());
+        }
+
+        let blueprints_cache: Blueprints = Arc::new(RwLock::new(bp_map));
```
Comment (folex) [inline particle-modules/src/modules.rs:95]: no need to specify type here

File: particle-modules/src/modules.rs [hunk 5]
```diff
@@ -79,10 +86,20 @@ impl ModuleRepository {
 
         let modules_by_name = Arc::new(Mutex::new(modules_by_name));
 
+        let blueprints = Self::load_blueprints(blueprints_dir);
+        let mut bp_map = HashMap::new();
+        for bp in blueprints.iter() {
+            bp_map.insert(bp.id.clone(), bp.clone());
+        }
```
Comment (folex) [inline particle-modules/src/modules.rs:93]: Maybe move that logic to `load_blueprints` function? Return HasMap right away

File: particle-modules/src/modules.rs [hunk 6]
```diff
@@ -195,18 +217,79 @@ impl ModuleRepository {
...
+            let lock = blueprints.read();
+            lock.get(id).cloned()
+        };
+
+        match bp {
+            None => Err(BlueprintNotFound { id: id.to_string() }),
+            Some(bp) => {
+                let dep = bp
+                    .get_facade_module()
+                    .ok_or(EmptyDependenciesList { id: id.to_string() })?;
```
Comment (folex) [inline particle-modules/src/modules.rs:233]: Hm. I wonder if we check that dependencies aren't empty on `add_blueprint`. Maybe lets check?

File: particle-modules/src/modules.rs [hunk 7]
```diff
@@ -195,18 +217,79 @@ impl ModuleRepository {
         })
     }
 
+    pub fn get_interface_by_blueprint_id(&self, id: &str) -> Result<JValue> {
```
Comment (folex) [inline particle-modules/src/modules.rs:220]:
```suggestion
    pub fn get_facade_interface(&self, blueprint_id: &str) -> Result<JValue> {
```

File: particle-modules/src/modules.rs [hunk 8]
```diff
@@ -195,18 +217,79 @@ impl ModuleRepository {
...
+    }
+
+    pub fn get_interface_by_hash(&self, hash: &Hash) -> Result<JValue> {
+        let modules_dir: PathBuf = self.modules_dir.clone();
+        let cache: Arc<RwLock<HashMap<Hash, JValue>>> = self.module_interface_cache.clone();
+
+        Self::get_interface_by_hash_internal(modules_dir, cache, hash)
+    }
+
+    fn get_interface_by_hash_internal(
```
Comment (folex) [inline particle-modules/src/modules.rs:252]: you can move that function from the `impl` block. Then you can call it without `Self::`

File: particle-modules/src/modules.rs [hunk 9]
```diff
@@ -195,18 +217,79 @@ impl ModuleRepository {
...
+            Some(interface) => interface,
+            None => {
+                let path = modules_dir.join(module_file_name(hash));
+
+                let interface = module_interface(&path)
+                    .map_err(|err| ReadModuleInterfaceError { path, err })?;
+
+                let json = json!(interface);
+
+                json
```
Comment (folex) [inline particle-modules/src/modules.rs:272]:
```suggestion
                let path = modules_dir.join(module_file_name(hash));
                let interface = module_interface(&path)
                    .map_err(|err| ReadModuleInterfaceError { path, err })?;
                let json = json!(interface);
                json
```

File: particle-modules/src/modules.rs [hunk 10]
```diff
@@ -218,46 +301,70 @@ impl ModuleRepository {
...
+                        Ok(blueprint) => Some(blueprint),
+                        Err(err) => {
+                            log::warn!("get_blueprints error on serialization: {:?}", err);
+                            None
                         }
-                    })
-                    .collect(),
-            ))
+                    }
+                })
```
Comment (folex) [inline particle-modules/src/modules.rs:360]: You can do `json!`, and that likely wouldn't require cloning & avoid error handling

File: particle-modules/src/modules.rs [hunk 11]
```diff
@@ -369,22 +472,48 @@ mod tests {
...
+                mounted_binaries: None,
+                logging_mask: None,
+            },
+        };
+
+        let hash = add_module(&repo, base64::encode(module), config).unwrap();
+
+        let result = get_interface(&repo, hash);
+        assert_eq!(0, result.ret_code);
+    }
```
Comment (folex) [inline particle-modules/src/modules.rs:518]: Maybe add a test that list of modules gets cached after get_interface? I believe you can access cache field from tests.

File: particle-services/src/app_services.rs [hunk 12]
```diff
@@ -424,6 +440,27 @@ mod tests {
         );
     }
 
+    #[test]
+    fn test_get_interface_cache() {
```
Comment (folex) [inline particle-services/src/app_services.rs:444]: How does it test cache?

File: particle-services/src/app_services.rs [hunk 13]
```diff
@@ -424,6 +440,27 @@ mod tests {
...
+        let local_pid = create_pid();
+        let management_pid = create_pid();
+        let base_dir = TempDir::new("test").unwrap();
+        let pas = create_pas(local_pid, management_pid, base_dir.path().into());
+
+        let module = load_module("../particle-node/tests/tetraplets/artifacts", "tetraplets");
+
+        let service_id1 = create_service(&pas, "tetra".to_string(), module.clone()).unwrap();
+        let service_id2 = create_service(&pas, "tetra".to_string(), module.clone()).unwrap();
```
Comment (folex) [inline particle-services/src/app_services.rs:453]: Maybe remove wasm module from filesystem to check that it is cached?

**Review by folex — CHANGES_REQUESTED** [review]

File: particle-modules/src/modules.rs [hunk 14]
```diff
@@ -42,12 +45,16 @@ pub struct AddBlueprint {
...
+type Blueprints = Arc<RwLock<HashMap<String, Blueprint>>>;
+
 #[derive(Clone)]
 pub struct ModuleRepository {
     modules_dir: PathBuf,
     blueprints_dir: PathBuf,
     /// Map of module_config.name to blake3::hash(module bytes)
     modules_by_name: Arc<Mutex<HashMap<ModuleName, Hash>>>,
+    module_interface_cache: Arc<RwLock<HashMap<Hash, JValue>>>,
+    blueprints: Blueprints,
```
Comment (folex) [inline particle-modules/src/modules.rs:57]: Discussed in slack and decided to remove that alias for consistency

File: particle-modules/src/modules.rs [hunk 15]
```diff
@@ -161,6 +154,12 @@ impl ModuleRepository {
                 .map(|module| Ok(Hash(resolve_hash(&modules, module)?)))
                 .collect::<Result<_>>()?;
 
+            if dependencies.is_empty() {
```
Comment (folex) [inline particle-modules/src/modules.rs:157]: I think that should go before `resolve_hash` for performance reasons

File: particle-modules/src/modules.rs [hunk 16]
```diff
@@ -381,6 +349,32 @@ impl ModuleRepository {
     }
 }
 
+fn get_interface_by_hash_internal(
```
Comment (folex) [inline particle-modules/src/modules.rs:352]: let's drop `_internal`

File: particle-services/src/app_services.rs [hunk 17]
```diff
@@ -454,6 +457,13 @@ mod tests {
         let service_id3 = create_service(&pas, "tetra".to_string(), module).unwrap();
 
         let inter1 = get_interface(&pas, service_id1);
+
+        // delete module and check that interfaces will be returned anyway
+        let dir = modules_dir(base_dir.path().into());
+        for entry in fs::read_dir(dir).unwrap() {
+            remove_file(entry.unwrap().path()).unwrap();
+        }
+
```
Comment (folex) [inline particle-services/src/app_services.rs:466]: I think it is better to check that expected module was really removed. For example, check that it was there after upload, and then it is missing after remove_file

**Review by folex — CHANGES_REQUESTED** [review]

**Review by folex — APPROVED** [review]

**Pull request merged**